### ISSUE TITLE
BIG-IP requesting reboot after declaration - should reboot if required

### ISSUE THREAD

**Issue by jgruber** [F5Networks/f5-declarative-onboarding#17]
After declaring:

```
{
    "schemaVersion": "1.0.0",
    "class": "Device",
    "async": true,
    "label": "Basic onboarding",
    "Common": {
        "class": "Tenant",
        "hostname": "test-asm.example.openstack.com",
        "myLicense": {
            "class": "License",
            "licenseType": "licensePool",
            "bigIqHost": "192.168.245.114",
            "bigIqUsername": "admin",
            "bigIqPassword": "XXXXXXXXX",
            "licensePool": "BIGIPVEREGKEYS",
            "reachable": "true",
            "bigIpUsername": "admin",
            "bigIpPassword": "XXXXXXXXX"
        },
        "myDns": {
            "class": "DNS",
            "nameServers": [
                "8.8.8.8",
                "2001:4860:4860::8844"
            ],
            "search": [
                "example.openstack.com"
            ]
        },
        "myNtp": {
            "class": "NTP",
            "servers": [
                "0.pool.ntp.org",
                "1.pool.ntp.org",
                "2.pool.ntp.org"
            ],
            "timezone": "UTC"
        },
        "myProvisioning": {
            "class": "Provision",
            "ltm": "nominal",
            "asm": "nominal"
        },
        "internal": {
            "class": "VLAN",
            "tag": 4093,
            "mtu": "{{internal_net_mtu}}",
            "interfaces": [
                {
                    "name": "1.2",
                    "tagged": false
                }
            ]
        },
        "internal-self": {
            "class": "SelfIp",
            "address": "{{internal_fixed_ip}}/{{internal_subnet_maskbits}}",
            "vlan": "internal",
            "allowService": "default",
            "trafficGroup": "traffic-group-local-only"
        },
        "external": {
            "class": "VLAN",
            "tag": 4094,
            "mtu": "{{external_net_mtu}}",
            "interfaces": [
                {
                    "name": "1.3",
                    "tagged": true
                }
            ]
        },
        "external-self": {
            "class": "SelfIp",
            "address": "{{external_fixed_ip}}/{{external_subnet_maskbits}}",
            "vlan": "external",
            "allowService": "none",
            "trafficGroup": "traffic-group-local-only"
        },
        "default": {
            "class": "Route",
            "gw": "{{external_subnet_gateway}}",
            "network": "default",
            "mtu": 1500
        },
        "dbvars": {
        	"class": "DbVariables",
        	"ui.advisory.enabled": true,
        	"ui.advisory.color": "orange",
        	"ui.advisory.text": "This device is under centralized management."
        }
    }
}
```

The PS1 prompt is: REBOOT REQUIRED

The XUI banner reads: 'The system must be rebooted for some changes to take effect.'

D0 version:  1.3.0-4
BIGIP-13.1.1-0.0.4.ALL_1SLOT

If there are any licensing + provisioning use cases which require reboot, the declarer should have the ability to say the reboot happens automatically, as the BIG-IP is not functional with their declaration until they reboot.

**Comment by jgruber** [comment]
In addition, the advisory banner is not enabled after reboot.

```
# tmsh list sys db ui.*
...
sys db ui.advisory.color {
    value "green"
}
sys db ui.advisory.enabled {
    value "false"
}
sys db ui.advisory.text {
    value "This is a system advisory message."
}
...
```

So there are other declared states which are not persisted through the reboot. The declared state should be the end state, even after reboot.

**Comment by seattlevine** [comment]
DO should have rebooted the system. The declarer does not need to indicate this, it is supposed to happen automatically.

Tracking with internal issue 249

**Comment by jgruber** [comment]
I had to go check restnoded.log on to see what was going on.. turns out there was an error in the declaration for a non-delcared VLAN (added by bootstrap script):

"01071412:3: Cannot delete IP (1.1.1.122) because it is used by the system config-sync setting."

I will add the HA VLAN and IP to the declaration to remove this specific use case, but it is an error that the config-sync config was not removed when a SelfIP is removed. I would assume we are not checking for config-sync, failover, and mirroring use of a SelfIP we are removing.

**Comment by seattlevine** [comment]
> I had to go check restnoded.log on to see what was going on

Was the error not reported in the response from DO?

**Comment by jgruber** [comment]
Sorry for the delay.. deadlines.. I will have to reproduce again, the whole things .. including DO container goes away each test run.

But I'm not wrong to say that if DO decided (in my case because I did not declare the Self IP) to remove a Self IP, it should remove it from config-sync, failover, and mirroring settings? Seems those are 'global' device level configurations.. so they belong in DO.

**Comment by seattlevine** [comment]
The theory behind DO is to record the state of the machine the first time you run DO but before any config changes are made. If you leave something out of the declaration, DO will try to restore the missing item to what was there when you first ran it. So, the answer is, it depends on what that initial config was and what your declaration is. One can argue that this is not declarative, but for some fields, DO needs something to restore to if the item is not in the declaration, so it tries to do that for everything it owns. We could have a discussion on whether we should modify DO so that it only restores to original config items that it has to. Items like default route, dns and ntp (which vary based on your environment), etc.

**Comment by jgruber** [comment]
Truth is... most of the time I will NEVER use DO to re-POST except to revoke a license. In most of my use cases, once the license is revoked, you simply delete the VE instance. Who cares about the configuration at this point.

So you are telling me that if I had declared a configsync or failoverAddress and then removed it, that would have properly deprovisioned that and remove the selfip properly? If so, let's close it.

I've given up on trying to convince anyone of the difference between an initial configuration, globals, and what DO does.

**Comment by seattlevine** [comment]
Resolved in 1.3.1
